SU2_CFD/src/integration/CMultiGridIntegration.cpp

```diff
@@ -0,0 +1,763 @@
+  unsigned long Point_Fine, Point_Coarse, iVertex;
+  unsigned short Boundary, iMarker, iChildren, iVar;
+  su2double Area_Parent, Area_Children;
+  const su2double *Solution_Fine = nullptr, *Solution_Coarse = nullptr;
+
+  const unsigned short nVar = sol_coarse->GetnVar();
+
+  su2double *Solution = new su2double[nVar];
+
+  SU2_OMP_FOR_STAT(roundUpDiv(geo_coarse->GetnPointDomain(), omp_get_num_threads()))
+  for (Point_Coarse = 0; Point_Coarse < geo_coarse->GetnPointDomain(); Point_Coarse++) {
+
+    Area_Parent = geo_coarse->node[Point_Coarse]->GetVolume();
+
+    for (iVar = 0; iVar < nVar; iVar++) Solution[iVar] = 0.0;
+
+    for (iChildren = 0; iChildren < geo_coarse->node[Point_Coarse]->GetnChildren_CV(); iChildren++) {
+      Point_Fine = geo_coarse->node[Point_Coarse]->GetChildren_CV(iChildren);
+      Area_Children = geo_fine->node[Point_Fine]->GetVolume();
+      Solution_Fine = sol_fine->GetNodes()->GetSolution(Point_Fine);
+      for (iVar = 0; iVar < nVar; iVar++)
+        Solution[iVar] -= Solution_Fine[iVar]*Area_Children/Area_Parent;
+    }
+
+    Solution_Coarse = sol_coarse->GetNodes()->GetSolution(Point_Coarse);
+
+    for (iVar = 0; iVar < nVar; iVar++)
+      Solution[iVar] += Solution_Coarse[iVar];
+
+    for (iVar = 0; iVar < nVar; iVar++)
+      sol_coarse->GetNodes()->SetSolution_Old(Point_Coarse,Solution);
+  }
+
+  delete [] Solution;
+
+  /*--- Remove any contributions from no-slip walls. ---*/
+
+  for (iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++) {
+    Boundary = config->GetMarker_All_KindBC(iMarker);
+    if ((Boundary == HEAT_FLUX) ||
+        (Boundary == ISOTHERMAL) ||
+        (Boundary == CHT_WALL_INTERFACE)) {
+
+      SU2_OMP_FOR_STAT(32)
+      for (iVertex = 0; iVertex < geo_coarse->nVertex[iMarker]; iVertex++) {
+
+        Point_Coarse = geo_coarse->vertex[iMarker][iVertex]->GetNode();
+
+        /*--- For dirichlet boundary condtions, set the correction to zero.
+         Note that Solution_Old stores the correction not the actual value ---*/
+
+        sol_coarse->GetNodes()->SetVelSolutionOldZero(Point_Coarse);
+
+      }
+    }
+  }
+
+  /*--- MPI the set solution old ---*/
+
+  SU2_OMP_MASTER
+  {
+    sol_coarse->InitiateComms(geo_coarse, config, SOLUTION_OLD);
+    sol_coarse->CompleteComms(geo_coarse, config, SOLUTION_OLD);
+  }
+  SU2_OMP_BARRIER
+
+  /// TODO: Need to check for possible race condition here (multiple coarse points setting the same fine).
+
+  SU2_OMP_FOR_STAT(roundUpDiv(geo_coarse->GetnPointDomain(), omp_get_num_threads()))
+  for (Point_Coarse = 0; Point_Coarse < geo_coarse->GetnPointDomain(); Point_Coarse++) {
+    for (iChildren = 0; iChildren < geo_coarse->node[Point_Coarse]->GetnChildren_CV(); iChildren++) {
+      Point_Fine = geo_coarse->node[Point_Coarse]->GetChildren_CV(iChildren);
+      sol_fine->LinSysRes.SetBlock(Point_Fine, sol_coarse->GetNodes()->GetSolution_Old(Point_Coarse));
+    }
+  }
```

Comment:
You might know the answer to this by heart, is there possibility for overlap in the multigrid agglomeration? i.e. one fine cell assigned to more than one coarse cell.
Empirically it seems the answer is no (as the code does not crash).